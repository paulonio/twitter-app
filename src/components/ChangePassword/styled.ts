import { styled } from 'styled-components';

export const PasswordForm = styled.form`
  display: flex;
  flex-direction: column;
  align-items: end;
  justify-content: center;
`;

export const ButtonWrapper = styled.div`
  margin-top: 10px;
  width: 30%;
`;
