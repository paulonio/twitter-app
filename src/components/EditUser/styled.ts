import { styled } from 'styled-components';

export const EditForm = styled.form`
  width: 50%;
  display: flex;
  flex-direction: column;
  align-items: start;
  justify-content: center;
`;

export const RadioLabel = styled.label`
  display: flex;
  align-items: center;
  gap: 8px;
`;

export const ControlsWrapper = styled.div`
  margin-top: 10px;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const ButtonWrapper = styled.div`
  width: 30%;
`;
