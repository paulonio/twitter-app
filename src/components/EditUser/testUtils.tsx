import React, { FC } from 'react';
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';

import store, { persistor } from '@src/store';

import Theme from '@components/Theme/Theme';
import { BrowserRouter } from 'react-router-dom';

interface CustomProviderProps {
  children: JSX.Element | JSX.Element[] | string | string[];
}

const CustomProvider: FC<CustomProviderProps> = ({ children }) => {
  return (
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <BrowserRouter>
          <Theme>{children}</Theme>
        </BrowserRouter>
      </PersistGate>
    </Provider>
  );
};

export default CustomProvider;
