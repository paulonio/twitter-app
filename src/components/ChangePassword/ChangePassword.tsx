import React from 'react';
import { SubmitHandler, useForm } from 'react-hook-form';
import { EmailAuthProvider, reauthenticateWithCredential, updatePassword } from 'firebase/auth';
import Button from '../Button/Button';
import { Input } from '../SignUp/styled';
import { PasswordForm, ButtonWrapper } from './styled';
import { auth } from '../../../firebase';

interface ChangePasswordForm {
  oldPassword: string;
  newPassword: string;
}

const updatePasswordRequest = async (newPassword: string, oldPassword: string) => {
  const user = auth.currentUser;
  console.log(user);
  try {
    if (user && user.email) {
      const creds = EmailAuthProvider.credential(user?.email, oldPassword);
      await reauthenticateWithCredential(user, creds);
      await updatePassword(user, newPassword);
      await auth.signOut();
    }
  } catch (error) {
    console.log(error);
  }
};

const ChangePassword = () => {
  const { register, handleSubmit } = useForm<ChangePasswordForm>();

  const onSubmit: SubmitHandler<ChangePasswordForm> = (data) => {
    updatePasswordRequest(data.newPassword, data.oldPassword);
  };

  return (
    <PasswordForm onSubmit={handleSubmit(onSubmit)}>
      <Input type="password" placeholder="Old password" {...register('oldPassword')} />
      <Input type="password" placeholder="New password" {...register('newPassword')} />
      <ButtonWrapper>
        <Button buttonType="primary">Change password</Button>
      </ButtonWrapper>
    </PasswordForm>
  );
};

export default ChangePassword;
