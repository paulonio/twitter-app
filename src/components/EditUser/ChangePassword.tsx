import React from 'react';
import { useDispatch } from 'react-redux';
import { SubmitHandler, useForm } from 'react-hook-form';

import { Input } from '@components/SignUp/styled';
import Button from '@components/Button/Button';

import { changePasswordRequest } from '@store/actions/actions';

import { PasswordForm, ButtonWrapper } from './styled';

export interface ChangePasswordForm {
  oldPassword: string;
  newPassword: string;
}

const ChangePassword = () => {
  const { register, handleSubmit } = useForm<ChangePasswordForm>();
  const dispatch = useDispatch();

  const onSubmit: SubmitHandler<ChangePasswordForm> = (data) => {
    dispatch(changePasswordRequest(data));
  };

  return (
    <PasswordForm onSubmit={handleSubmit(onSubmit)}>
      <Input
        type="password"
        placeholder="Old password"
        {...register('oldPassword')}
        data-testid="password-input"
      />
      <Input type="password" placeholder="New password" {...register('newPassword')} />
      <ButtonWrapper>
        <Button $buttonType="primary">Change password</Button>
      </ButtonWrapper>
    </PasswordForm>
  );
};

export default ChangePassword;
