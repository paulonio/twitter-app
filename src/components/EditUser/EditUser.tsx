import React, { FC } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { SubmitHandler, useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';

import Modal from '@components/Modal/Modal';
import { Input } from '@components/SignUp/styled';
import ChangePassword from '@components/EditUser/ChangePassword';
import Button from '@components/Button/Button';

import { logoutRequest, updateUserRequest } from '@store/actions/actions';
import type { UserType } from '@store/slices/authSlice';
import type { StoreType } from '@store/index.ts';

import { ButtonWrapper, EditForm, RadioLabel, ControlsWrapper } from './styled';

interface EditUserProps {
  isActive: boolean;
  setActive: (value: boolean) => void;
}

export interface EditUserForm {
  name: string;
  lastName: string;
  gender: 'male' | 'female';
  telegram: string;
}

const schema = yup
  .object({
    name: yup.string().trim().min(3).required(),
    lastName: yup.string().trim(),
    telegram: yup.string().trim(),
  })
  .required();

const EditUser: FC<EditUserProps> = ({ isActive, setActive }) => {
  const user = useSelector<StoreType, UserType | null>((state) => state.auth.user);
  const { register, handleSubmit } = useForm<EditUserForm>({
    resolver: yupResolver(schema),
    defaultValues: {
      name: user?.displayName.split(' ')[0],
      lastName: user?.displayName.split(' ')[1],
      gender: user?.gender,
      telegram: user?.telegram,
    },
  });
  const dispatch = useDispatch();

  const onSubmit: SubmitHandler<EditUserForm> = (data) => {
    dispatch(updateUserRequest(data));
  };

  return (
    <Modal isActive={isActive} setActive={setActive}>
      <EditForm onSubmit={handleSubmit(onSubmit)}>
        <Input
          type="text"
          placeholder="Write your name"
          {...register('name')}
          data-testid="input-name"
        />
        <Input type="text" placeholder="Write your last name" {...register('lastName')} />
        <Input type="text" placeholder="Your telegram link" {...register('telegram')} />
        <ControlsWrapper>
          <RadioLabel>
            <Input
              id="male"
              type="radio"
              value="male"
              {...register('gender')}
              data-testid="radio-button"
            />
            Male
          </RadioLabel>
          <RadioLabel>
            <Input id="female" type="radio" value="female" {...register('gender')} />
            Female
          </RadioLabel>
        </ControlsWrapper>
        <ButtonWrapper>
          <Button $buttonType="primary">Submit</Button>
        </ButtonWrapper>
      </EditForm>
      <button type="button" onClick={() => dispatch(logoutRequest())}>
        Log out
      </button>
      <ChangePassword />
    </Modal>
  );
};

export default EditUser;
