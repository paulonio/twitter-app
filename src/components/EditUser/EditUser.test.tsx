import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import EditUser from '@components/EditUser/EditUser';
import ChangePassword from '@components/EditUser/ChangePassword';
import CustomProvider from './testUtils';

describe('Edit user', () => {
  it('should render edit user component', () => {
    const { container } = render(
      <CustomProvider>
        <EditUser isActive setActive={() => {}} />
      </CustomProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render change password component', () => {
    const { container } = render(
      <CustomProvider>
        <ChangePassword />
      </CustomProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('sholud change inputs correctly', () => {
    render(
      <CustomProvider>
        <EditUser isActive setActive={() => {}} />
      </CustomProvider>
    );

    const input: HTMLInputElement = screen.getByTestId('input-name');
    fireEvent.change(input, { target: { value: 'Test' } });
    expect(input.value).toEqual('Test');
  });

  it('should change radio button', () => {
    render(
      <CustomProvider>
        <EditUser isActive setActive={() => {}} />
      </CustomProvider>
    );

    const radio: HTMLInputElement = screen.getByTestId('radio-button');
    expect(radio.checked).toEqual(false);
    fireEvent.click(radio);
    expect(radio.checked).toEqual(true);
  });

  it('should change input in change password form', () => {
    render(
      <CustomProvider>
        <ChangePassword />
      </CustomProvider>
    );

    const input: HTMLInputElement = screen.getByTestId('password-input');
    fireEvent.change(input, { target: { value: 'Test' } });
    expect(input.value).toEqual('Test');
  });
});
