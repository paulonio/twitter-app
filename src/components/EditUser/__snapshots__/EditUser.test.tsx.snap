// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit user should render change password component 1`] = `
<div>
  <form
    class="sc-burPHO QhFRB"
  >
    <input
      class="sc-hcxXMm cQbYlz"
      data-testid="password-input"
      name="oldPassword"
      placeholder="Old password"
      type="password"
    />
    <input
      class="sc-hcxXMm cQbYlz"
      name="newPassword"
      placeholder="New password"
      type="password"
    />
    <div
      class="sc-jySWPz kmkKzT"
    >
      <button
        class="sc-SYvCl bqLUHW"
      >
        Change password
      </button>
    </div>
  </form>
</div>
`;

exports[`Edit user should render edit user component 1`] = `<div />`;
